src/model/prompt_mask.rs

```diff
@@ -0,0 +1,55 @@
+use mlua::{Integer as LuaInt, String as LuaString, Table as LuaTable};
+use std::collections::BTreeMap;
+
+#[derive(Debug, PartialEq, Eq, Clone)]
+pub struct PromptMask {
+    mask: BTreeMap<i32, String>,
+}
+
+impl PromptMask {
+    pub fn new() -> Self {
+        PromptMask {
+            mask: BTreeMap::new(),
+        }
+    }
+
+    pub fn clear(&mut self) {
+        self.mask.clear()
+    }
+
+    pub fn mask_buffer(&self, buf: &[char]) -> String {
+        let mut masked_buf = buf.to_owned();
+        let mut offset = 0;
+        for (idx, mask) in self.mask.iter() {
+            // NB: idx subtracted by one to account for Lua one-indexing.
+            let adjusted_idx = offset + (idx - 1) as usize;
+
+            // TODO(XXX): This is a HACK! Our mask/buf are falling out of sync somehow...
+            if adjusted_idx > masked_buf.len() {
+                return masked_buf.iter().collect::<String>();
+            }
```

Comment:
I mention this in the WIP PR's TODO list. I've noticed if I enter some data with a prompt mask, then send the data, then reload the data from history, then press "down" to clear the input prompt, something weird happens:

* Pressing the down arrow ends up setting the prompt content to a new empty buffer.
* We clear the prompt mask when that happens
* But! Somehow `self.mask.iter` on L23 returns the old mask content, even though the mask has been cleared.
* This results in the `adjusted_idx` being outside the bounds of the `masked_buf` (which is empty), and a panic ensues.

I added this bounds check here as a temporary hack and it "solves" the problem. Maybe someone more familiar with the project can spot my error?